### DISCUSSION TITLE
Is a C/C++ compiler required?

### DISCUSSION THREAD

**Discussion by giuseppe-g-gelardi** [nvim-treesitter/nvim-treesitter#4497] [General]
Hi, Sorry if this a dumb question.

Is a C compiler required? Trying to get neovim to work on my work computer but its super locked down and im unable to really do too much outside of install vscode and do the tickets im assigned. found a workaround for some things but am unable to install a C compiler ergo am unable to install any parsers. 

is it possible to clone, say the tree-sitter-typescript and tree-sitter-go repos manually and it work or am i totally boned here?

thank you for any help.

**Comment by clason** [comment]
Yes. (Sorry, there's no two ways around this; treesitter currently has no binary distribution strategy. So the parsers need to be compiled on the host system.)

**Comment by amaanq** [comment]
You need a C compiler, and a few parsers require npm. I find it really weird that you can't install a **compiler** on a laptop, only solution beyond having a compiler is compiling on another machine and copying it over (make sure you compile for the right os/arch)

**Comment by clason** [comment]
That being said, if you have a non-locked down system (of the same architecture etc.) you can use to install the parsers for, you can just put the.so files and queries on a thumbstick and copy them to your config directory.

**Reply by amaanq** [reply]
or .dll if it's *windows*

**Reply by giuseppe-g-gelardi** [reply]
> or .dll if it's _windows_

its windows. one of the errors i kept getting was something about .so files.  i just checked my personal machine, also windows 10 and all of the parsers are .so as well. unless im looking in the wrong place, which is entirely possible

**Reply by giuseppe-g-gelardi** [reply]
> That being said, if you have a non-locked down system (of the same architecture etc.) you can use to install the parsers for, you can just put the.so files and queries on a thumbstick and copy them to your config directory.

forgive me if this seems like an elementary question but..

everything in ~$HOME/nvim-data/lazy/nvim-treesitter/parser dir on a thumbstick? or would it be better to do the whome nvim-treesitter dir?

**Reply by amaanq** [reply]
Tbh idk how neovim works on windows at all so maybe somehow it runs .sos there too. 

> everything in ~$HOME/nvim-data/lazy/nvim-treesitter/parser dir on a thumbstick? or would it be better to do the whome nvim-treesitter dir?

Just the parsers, aka all the .so/.dll binaries, and then throw them into the parser dir on your locked down machine

**Reply by giuseppe-g-gelardi** [reply]
> Tbh idk how neovim works on windows at all so maybe somehow it runs .sos there too. 
> 
> 
> 
> > everything in ~$HOME/nvim-data/lazy/nvim-treesitter/parser dir on a thumbstick? or would it be better to do the whome nvim-treesitter dir?
> 
> 
> 
> Just the parsers, aka all the .so/.dll binaries, and then throw them into the parser dir on your locked down machine

Thank you, I'll give it a try.

**Comment by LatiaPensingervS** [comment]
No worries, it's not a dumb question at all!

In short, yes, a C compiler is required to build and install parsers such as tree-sitter. This is because tree-sitter is implemented in C and needs to be compiled into machine code that can be run on your computer. So unfortunately, if you don't have a C compiler installed, you won't be able to build and install tree-sitter or any other software that requires compilation.

That being said, there may be some workarounds or alternatives you can explore. One option is to see if your company has any sort of software management system or package manager that allows you to install pre-built software packages without needing to compile them yourself. For example, if your company uses a Linux distribution such as Red Hat or CentOS, you may be able to use the yum package manager to install pre-built versions of tree-sitter or other tools.

Another option is to see if you can use an online C compiler or development environment. There are several websites and online services that allow you to write and compile C code directly in your web browser, without needing to install any software locally. This could potentially allow you to build and install tree-sitter or other software without needing a C compiler installed on your local machine.

Finally, if none of these options work for you, you may need to consider using a different tool or approach that doesn't require compilation. Depending on your specific needs, there may be other parsers or plugins available that are written in a different language or don't require compilation.

**Reply by giuseppe-g-gelardi** [reply]
> No worries, it's not a dumb question at all!
> 
> In short, yes, a C compiler is required to build and install parsers such as tree-sitter. This is because tree-sitter is implemented in C and needs to be compiled into machine code that can be run on your computer. So unfortunately, if you don't have a C compiler installed, you won't be able to build and install tree-sitter or any other software that requires compilation.
> 
> That being said, there may be some workarounds or alternatives you can explore. One option is to see if your company has any sort of software management system or package manager that allows you to install pre-built software packages without needing to compile them yourself. For example, if your company uses a Linux distribution such as Red Hat or CentOS, you may be able to use the yum package manager to install pre-built versions of tree-sitter or other tools.
> 
> Another option is to see if you can use an online C compiler or development environment. There are several websites and online services that allow you to write and compile C code directly in your web browser, without needing to install any software locally. This could potentially allow you to build and install tree-sitter or other software without needing a C compiler installed on your local machine.
> 
> Finally, if none of these options work for you, you may need to consider using a different tool or approach that doesn't require compilation. Depending on your specific needs, there may be other parsers or plugins available that are written in a different language or don't require compilation.

thank you for the reply!

there is not too much i can do, its a locked down windows 10 machine. i did, however, take all the .so files from my personal (windows 10) machine, put them in a repo and clone them at work and that gave me syntax highlighting but not diagnostics or anything else. i think the issue is its typescript.

i went through this: https://github.com/nvim-treesitter/nvim-treesitter/wiki/Windows-support and did all i could. my request to install visual studio was denied. 

that being said, again sorry for the incoming dumb question. how would i use an online C compiler to work with treesitter? i think i might be able to get away with that.

**Reply by amaanq** [reply]
tree-sitter doesn't provide diagnostics, you'll need an LSP for that.

**Reply by giuseppe-g-gelardi** [reply]
> tree-sitter doesn't provide diagnostics, you'll need an LSP for that. 

My mistake

**Comment by GroverHyrex** [comment]
Hi there,

It's not a dumb question at all. A C compiler is typically required to build Neovim from source, but you may be able to use a pre-built binary release instead. As for your specific situation, cloning the tree-sitter-typescript and tree-sitter-go repos manually may work, but it depends on how Neovim is set up to use those parsers. You could try building the parsers on a different machine with a C compiler, and then transferring the resulting binaries to your work computer. Alternatively, you may be able to find pre-built binaries of the parsers online and use those instead. Good luck!

**Reply by amaanq** [reply]
I'd advise going around online hunting for one, there likely isn't a place to have it that is trustworthy - the best solution is compiling it on a different machine and transferring it over.

And fwiw, we've wanted to have a central registry for compiled parsers for all major platforms/systems for a while now, tsk tsk

**Reply by giuseppe-g-gelardi** [reply]
> Hi there,
> 
> 
> 
> It's not a dumb question at all. A C compiler is typically required to build Neovim from source, but you may be able to use a pre-built binary release instead. As for your specific situation, cloning the tree-sitter-typescript and tree-sitter-go repos manually may work, but it depends on how Neovim is set up to use those parsers. You could try building the parsers on a different machine with a C compiler, and then transferring the resulting binaries to your work computer. Alternatively, you may be able to find pre-built binaries of the parsers online and use those instead. Good luck!

I actually ended up adding all of the compiled parsers to my repo and cloning it down and it mostly worked. I found out Typescript/tsx was the only one giving me an issue. 

Thank you !

**Reply by giuseppe-g-gelardi** [reply]
> I'd advise going around online hunting for one, there likely isn't a place to have it that is trustworthy - the best solution is compiling it on a different machine and transferring it over.
> 
> 
> 
> And fwiw, we've wanted to have a central registry for compiled parsers for all major platforms/systems for a while now, tsk tsk

That would be cool. I'd offer to help but I'm well out of my depth in general haha.

**Comment by LazaroPirnero** [comment]
In order to compile C code, you do need a C compiler installed on your computer. Without it, you won't be able to compile the code and install the parsers.

Regarding your specific situation, it's possible that you might be able to manually clone the repositories for tree-sitter-typescript and tree-sitter-go and get them to work, but it's hard to say for sure without more information about your setup and the specific requirements of those parsers.

If you're unable to install a C compiler on your work computer due to restrictions, you might want to consider using an online compiler like CodeChef, Compiler Explorer, or repl.it. These platforms allow you to compile and run code in a web browser without requiring any local installations.

Alternatively, you could try to see if there are any pre-built binaries available for the parsers you need that don't require compiling. Some repositories might include pre-built binaries in their releases or as part of their installation instructions.

**Comment by theHamsta** [comment]
I plan to provide binaries as a external project. We previously also uploaded artifacts from our CI, but this would require you a re-download whenever nvim-treesitter update. Also see https://github.com/nvim-treesitter/nvim-treesitter/pull/3688

**Comment by HipolitoAkram4Fr** [comment]
a C compiler is required for installing and building parsers. If your work computer is super locked down and you can't install a C compiler, then you might be out of luck here.

As for your second question, manually cloning the repos might work, but it really depends on the specific setup and environment you're working with. It might be worth giving it a try and seeing if it works for you.
